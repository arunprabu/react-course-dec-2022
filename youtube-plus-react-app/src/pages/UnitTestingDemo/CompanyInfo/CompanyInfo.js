import { PropTypes } from 'prop-types';
import React, { useState } from 'react';

const CompanyInfo = ( props ) => {
  const [countryName, setCountryName] = useState('USA');

  return (
    <div>
      <h2>CompanyInfo | Testing Props, Styles, Placeholder Text, Events, States, Snapshot</h2>
      <h3 data-testid="companyName">Company Name: Cognizant</h3>
      <p data-testid="foundedYear" style={{ color: 'rgb(0, 255, 0)' }}>
        Founded In: {props.foundedYear}
      </p>
      <input
        type="text"
        placeholder="Enter Your Country Name"
        value={countryName}
        onChange={(event) => {
          setCountryName(event.target.value);
        }}
      />
      <p data-testid="visitCognizant">Visit Cognizant {countryName} Website</p>
    </div>
  );
};
CompanyInfo.propTypes = {
  foundedYear: PropTypes.string
};

export default CompanyInfo;
