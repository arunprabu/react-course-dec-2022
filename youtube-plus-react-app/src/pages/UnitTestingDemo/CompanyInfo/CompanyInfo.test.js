import { fireEvent, render, screen } from '@testing-library/react';
import CompanyInfo from './CompanyInfo';

// test suite == group of related test specs
describe('CompanyInfo', () => {
  
  it('should have companyName Cognizant', () => {
    render(<CompanyInfo />);
    const companyNameElement = screen.getByTestId('companyName');
    console.log(companyNameElement.textContent);
    expect(companyNameElement.textContent).toBe('Company Name: Cognizant');
  });

  // testing props -- must
  it('receives foundedYear prop and displays in JSX', () => {
    // when you are testing prop, render the comp with necessary props
    render(<CompanyInfo foundedYear="2006"/>);
    const foundedYearElement = screen.getByTestId('foundedYear');
    expect(foundedYearElement).toHaveTextContent('Founded In: 2006');
  });

  // TODO: test whether the comp is receiving a prop 'employeeCount' 
  
  // form input with placeholder
  it('has an input with \'Enter Your Country Name\' as placeholder text', () => {
    render(<CompanyInfo />);
    expect(screen.getByPlaceholderText('Enter Your Country Name')).toBeTruthy();
  });

  // inline styles 
  it('has foundedYear with text color green -- inline styles', () => {
    render(<CompanyInfo foundedYear="2006" />);
    const foundedYearElement = screen.getByTestId('foundedYear');
    // recommended to use color code in hexa or rgb -- not the color names
    expect(foundedYearElement).toHaveStyle('color: rgb(0, 255, 0)');
  });

  // TODO: test whether a button has css class btn-primary 
    // the button text should be Join Cognizant 
    // use data-testid 
    // you can have muliple expectations here 
  
  // events and states 
  it('displays countryName by default and updates UI onChange of input', () => {
    render(<CompanyInfo />);
    // find the input element
    const countryNameInput = screen.getByPlaceholderText('Enter Your Country Name');
    // now checking whether the onChange event handler is working or not
    const mockEventObj = {
      // praparing the mock event obj with target value being China
      target: {
        value: 'Japan'
      }
    };
    // let's trigger the event -- thru program
    fireEvent.change(countryNameInput, mockEventObj);
    // finally checking if the input element is showing the change
    expect(countryNameInput.value).toBe('Japan');

    // have expectation
    expect(screen.getByTestId('visitCognizant')).toHaveTextContent('Visit Cognizant Japan Website');
  });

  // snapshot test

  // mocks 
});