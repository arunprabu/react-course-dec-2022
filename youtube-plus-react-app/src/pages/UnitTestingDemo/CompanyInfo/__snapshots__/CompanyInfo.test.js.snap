// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CompanyInfo has right snapshot with all requirements completed 1`] = `
<div>
  <h2>
    CompanyInfo | Testing Props, Styles, Placeholder Text, Events, States and Snapshot
  </h2>
  <p>
    Wed Jan 18 2023 12:37:00 GMT+0530 (India Standard Time)
  </p>
  <h3
    data-testid="companyName"
  >
    Company Name: Cognizant
  </h3>
  <p
    data-testid="foundedYear"
    style={
      Object {
        "color": "rgb(0, 255, 0)",
      }
    }
  >
    Founded In: 
    2006
  </p>
  <input
    onChange={[Function]}
    placeholder="Enter Your Country Name"
    type="text"
    value="USA"
  />
  <p
    data-testid="visitCognizant"
  >
    Visit Cognizant 
    USA
     Website
  </p>
</div>
`;
